src/sqlfluff/rules/L054.py

```diff
@@ -0,0 +1,111 @@
+"""Implementation of Rule L054."""
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, RuleContext
+
+
+class Rule_L054(BaseRule):
+    """Inconsistent column references in GROUP BY/ORDER BY clauses.
+
+    | **Anti-pattern**
+    | A mix of implicit and explicit column references are used in a GROUP BY clause.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, 2;
+
+        -- The same also applies to column
+        -- references in ORDER BY clauses.
+
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, bar;
+
+    | **Best practice**
+    | Reference all GROUP BY/ORDER BY columns either by name or by position.
+
+    .. code-block:: sql
+       :force:
+
+        -- GROUP BY: Explicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            foo, bar;
+
+        -- ORDER BY: Explicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            foo, bar;
+
+        -- GROUP BY: Implicit
+        SELECT
+            foo,
+            bar,
+            sum(baz) AS sum_value
+        FROM fake_table
+        GROUP BY
+            1, 2;
+
+        -- ORDER BY: Implicit
+        SELECT
+            foo,
+            bar
+        FROM fake_table
+        ORDER BY
+            1, 2;
+    """
+
+    config_keywords = ["group_by_and_order_by_style"]
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Inconsistent column references in GROUP BY/ORDER BY clauses."""
+        # Config type hints
+        self.group_by_and_order_by_style: str
+
+        # We only care about GROUP BY/ORDER BY clauses.
+        if context.segment.name not in {"GroupByClauseSegment", "OrderByClauseSegment"}:
+            return None
+
+        # Look at child segments to detect the presence of disallowed segments.
+        child_segment_names = [segment.name for segment in context.segment.segments]
+        disallowed_segment_names = {
+            "explicit": "numeric_literal",
+            "implicit": "ColumnReferenceSegment",
+        }
+
+        if self.group_by_and_order_by_style == "consistent":
+            # If consistent naming then raise lint error if both
+            # numeric literal and column reference segments are present.
+            if all(
```

Comment:
I'm struggling to understand how this code works -- I _think_ I understand, but it seems like a pretty indirect way of doing it. Can you think of a clearer way to express this logic?

Reply:
sure it basically just says if the ORDER BY/GROUP BY clause contains both numeric and column reference values then it's inconsistent

Reply:
I thought it was pretty clear. And was one of the reasons for me "nice code" comment.

Can you explain more what you're struggling with @barrywhart ? Maybe I'm assuming something that's not true!

Reply:
It took me a while to figure out that we were checking if all segment name categories appeared in the list, that was wrong. The code is correct. It would read more clearly to me if it were using sets, and a set with size > 1 was an error.

This is not a big worry for me -- good to know that it's readable for you. I do like the functional style. 👍

Reply:
I've updated the code flow so should hopefully be a lot clearer now :)